src/main/java/com/squareup/kotlinpoet/TypeSpec.kt

```diff
@@ -423,18 +432,43 @@ class TypeSpec private constructor(builder: TypeSpec.Builder) {
     }
 
     fun addSuperinterfaces(superinterfaces: Iterable<TypeName>) = apply {
-      this.superinterfaces += superinterfaces
+      this.superinterfaces.putAll(superinterfaces.map {
+        Pair(it, null)
+      })
     }
 
-    fun addSuperinterface(superinterface: TypeName) = apply {
-      superinterfaces += superinterface
+    fun addSuperinterface(superinterface: TypeName, delegate: CodeBlock = CodeBlock.EMPTY) = apply {
+      if (delegate.isEmpty()) {
+        this.superinterfaces.put(superinterface, null)
+      } else {
+        require(kind == Kind.CLASS) {
+          "delegation only allowed for classes (found $kind '$name')" }
+        require(!superinterface.nullable) {
+          "expected non-nullable type but was '${superinterface.asNonNullable()}'"}
+        require(this.superinterfaces[superinterface] == null) {
+          "'$name' can not delegate to $superinterface by $delegate with existing declaration by " +
+              "${this.superinterfaces[superinterface]}" }
+        this.superinterfaces[superinterface] = delegate
+      }
     }
 
-    fun addSuperinterface(superinterface: Type)
-        = addSuperinterface(superinterface.asTypeName())
+    fun addSuperinterface(superinterface: Type, delegate: CodeBlock = CodeBlock.EMPTY)
+        = addSuperinterface(superinterface.asTypeName(), delegate)
+
+    fun addSuperinterface(superinterface: KClass<*>, delegate: CodeBlock = CodeBlock.EMPTY)
+        = addSuperinterface(superinterface.asTypeName(), delegate)
 
-    fun addSuperinterface(superinterface: KClass<*>)
-        = addSuperinterface(superinterface.asTypeName())
+    fun addSuperinterface(superinterface: KClass<*>, constructorParameterName: String) =
+        addSuperinterface(superinterface.asTypeName(), constructorParameterName)
+
+    fun addSuperinterface(superinterface: TypeName, constructorParameter: String) = apply {
+      requireNotNull(primaryConstructor) {
+        "delegating to constructor parameter requires not-null constructor" }
+      val parameter = primaryConstructor?.parameter(constructorParameter)
+      requireNotNull(parameter) {
+        "no such constructor parameter '$constructorParameter' to delegate to for type '$name'" }
+      addSuperinterface(superinterface, CodeBlock.of(constructorParameter))
```

Comment:
This is academic, but strictly speaking we shouldn’t use `CodeBlock.of()` to turn a literal name into a codeblock. The problem is that the literal name could have a metacharacter and we want to escape it.

Better would be `CodeBlock.of("%L", constructorParameter)`. That way if you pass in the _string_ `%[` that’s what we’ll echo out. As-is that’ll mess up our indentation and it’ll be difficult to diagnose why.